R/blup_analysis.R

```diff
@@ -0,0 +1,522 @@
+    }
+
+    lmm_model <- NULL # Ensure model is null if not assigned
+
+    tryCatch({
+      # Filter data for current concept
+      concept_data <- lab_data %>%
+        filter(.data$OMOP_CONCEPT_ID == concept_id)
+
+      # Apply drug exposure filter if requested
+      if (drug_exposed_only && !is.na(atc_code) && !is.null(all_drug_purchases)) {
+        # Get FINNGENIDs who purchased this specific drug
+        drug_exposed_ids <- all_drug_purchases %>%
+          filter(.data$ATC == atc_code) %>%
+          select(.data$FINNGENID) %>%
+          distinct() %>%
+          pull(.data$FINNGENID)
+
+        # Filter to only drug-exposed individuals with non-NA first_drug_age
+        if ("first_drug_age" %in% colnames(concept_data)) {
+          concept_data <- concept_data %>%
+            filter(.data$FINNGENID %in% drug_exposed_ids & !is.na(.data$first_drug_age))
+        } else {
+          concept_data <- concept_data %>%
+            filter(.data$FINNGENID %in% drug_exposed_ids)
+        }
+      }
+
+      # Count measurements per individual
+      measurement_counts <- concept_data %>%
+        group_by(.data$FINNGENID) %>%
+        summarise(n_measurements = n()) %>%
+        filter(.data$n_measurements >= min_measurements)
+
+      # Filter to individuals with sufficient measurements
+      analysis_data <- concept_data %>%
+        filter(.data$FINNGENID %in% measurement_counts$FINNGENID)
+
+      # Check if there's enough data
+      n_individuals <- length(unique(analysis_data$FINNGENID))
+      if (n_individuals < 10) {
+        warning(paste0("Only ", n_individuals, " individuals for OMOP_CONCEPT_ID ",
+                       concept_id, ". Skipping analysis."))
+        next
+      }
+
+      # Standardize variables to improve convergence
+      # Scale age by its SD
+      age_mean <- mean(analysis_data$EVENT_AGE, na.rm = TRUE)
+      age_sd <- sd(analysis_data$EVENT_AGE, na.rm = TRUE)
+      if (age_sd == 0) {
+        warning(paste0("No variation in age for OMOP_CONCEPT_ID ", concept_id, ". Skipping analysis."))
+        next
+      }
+      analysis_data$EVENT_AGE_SCALED <- (analysis_data$EVENT_AGE - age_mean) / age_sd
+
+      # Scale lab values by their SD
+      lab_mean <- mean(analysis_data$MEASUREMENT_VALUE_HARMONIZED, na.rm = TRUE)
+      lab_sd <- sd(analysis_data$MEASUREMENT_VALUE_HARMONIZED, na.rm = TRUE)
+      if (lab_sd == 0) {
+        warning(paste0("No variation in lab values for OMOP_CONCEPT_ID ", concept_id, ". Skipping analysis."))
+        next
+      }
+      analysis_data$LAB_VALUE_SCALED <- (analysis_data$MEASUREMENT_VALUE_HARMONIZED - lab_mean) / lab_sd
+
+      # Fit the linear mixed model
+      # Model: lab_value ~ sex + age + (age | FINNGENID)
+      # This includes random intercepts and random slopes for age
+      model_converged <- FALSE
+
+      # First try the full model with correlated random effects
+      if (include_sex) {
+        lmm_model <- lme4::lmer(
```

Comment:
could have formulat as string and if statement would append sex to that. Not the model parameters are unnecessarily repeated and easy to forget to change only on when making changes. less code is always better than more.